src/workspace.jl

```diff
@@ -1,48 +1,58 @@
-ismacro(f::Function) = startswith(string(methods(f).mt.name), "@")
-
-wstype(x) = ""
-wstype(::Module) = "module"
-wstype(f::Function) = "function"
-wstype(::Type) = "type"
-wstype(::Expr) = "mixin"
-wstype(::Symbol) = "tag"
-wstype(::AbstractString) = "property"
-wstype(::Number) = "constant"
-wstype(::Exception) = "tag"
+handle("workspace") do mod
+  mod = getmodule′(mod)
+  ns = Symbol.(CodeTools.filtervalid(names(mod; all = true)))
+  filter!(ns) do n
+    !Base.isdeprecated(mod, n) && isdefined(mod, n) && n != Symbol(mod)
+  end
+  contexts = [d(:context => string(mod), :items => map(n -> wsitem(mod, n), ns))]
+  isdebugging() && prepend!(contexts, JunoDebugger.contexts())
+  contexts
+end
 
-wsicon(x) = ""
-wsicon(f::Function) = ismacro(f) ? "icon-mention" : ""
-wsicon(::AbstractArray) = "icon-file-binary"
-wsicon(::AbstractVector) = "icon-list-ordered"
-wsicon(::AbstractString) = "icon-quote"
-wsicon(::Expr) = "icon-code"
-wsicon(::Symbol) = "icon-code"
-wsicon(::Exception) = "icon-bug"
-wsicon(::Number) = "n"
+wsitem(mod, name) = begin
+  val = getfield(mod, name)
+  wsitem(mod, name, val)
+end
+wsitem(mod, name, val) = begin
+  d(:name       => name,
+    :value      => render′(Inline(), val),
+    :nativetype => DocSeeker.determinetype(mod, name),
+    :type       => wstype(mod, name, val),
+    :icon       => wsicon(mod, name, val))
+end
 
-wsnamed(name, val) = false
-wsnamed(name, f::Function) = name == methods(f).mt.name
-wsnamed(name, m::Module) = name == module_name(m)
-wsnamed(name, T::DataType) = name == Symbol(T.name)
+#=
+@NOTE: `wstype` and `wsicon` are also used for completions / docs
+=#
 
-function wsitem(name, val)
-  d(:name  => name,
-    :value => render′(Inline(), val),
-    :type  => wstype(val),
-    :icon  => wsicon(val))
+wstype(mod, name, val) = begin
```

Comment:
This seems less clear than the previous dispatch based solution. What's the reason for the change?

Reply:
I agree with it's less clear and readable.

Why I changed into `isa` (`<:`) branching instead of dispatches is performance: dispatches are slower than explicit conditional branching when the arguments aren't concrete
(You can find my simple benchmarks here: https://gist.github.com/aviatesk/c8c0a24d30051151433f44431e7f54f1)

Of course dispatches automatically do type specializations but I'm fairly sure about the branching above works (even though I suspect we can improve the logic around: https://github.com/JunoLab/Atom.jl/blob/764dfc4aca8cdc61dc35521588cbd7fea3cc5f69/src/workspace.jl#L31-L33 especially)

Reply:
I.e.: I made a point of not doing something like:
```julia
isabranch() = begin
  if ary isa AbstractArray ...
  elseif ary isa AbstractVector ... # never reached
  else # fallback
  end
end
```

But maybe readability might more important here for us, thus I can change it into dispatch-based branching :)

Reply:
I just simplified the logic and I hope it makes more sense to you 👍